### PR TITLE
Change the maxDurationBeforeRetry value

### LINKED ISSUE #75230: CSI dirver that implemented attacher would not be called by kubelet to mount volume in time
**What happened**:

csi driver that implemented attacher  would not be called by kubelet to mount volume in time though it aleady attached the volume to a node.

**What you expected to happen**:

When a pv is attahced to a node, `mount` should be called in time.

**How to reproduce it (as minimally and precisely as possible)**:

1. Using an existing pv in the pod and the pV driver is a csi driver
2. The csi dirver(supporting attacher)

**Anything else we need to know?**:

image

kubelet started to attach/mount the volume, and ran into the exponentialBackOffError,  so when a pv already attached to a node, kubelet would not call `mount` in time.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.11 ~ latest
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by cwdsuzhou**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
exponentialBackOff would lead to driver implementing attacher not mounting volume in time

**Which issue(s) this PR fixes**:

Fixes #75230 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/volume/util/operationexecutor/operation_generator.go [hunk 1]
```diff
@@ -1310,11 +1310,11 @@ func (og *operationGenerator) GenerateVerifyControllerAttachedVolumeFunc(
...
-			// error. Caller will log and retry. The node status is updated
-			// periodically by kubelet, so it may take as much as 10 seconds
-			// before this clears.
-			// Issue #28141 to enable on demand status updates.
-			return volumeToMount.GenerateError("Volume has not been added to the list of VolumesInUse in the node's volume status", nil)
+			// nil. Kubelet will retry in the following loops.
+
+			// Issue #75230 exponentialBackOff would lead to
+			// driver implementing attacher not mounting volume in time.
+			return nil, nil
```
Comment (cwdsuzhou) [inline pkg/volume/util/operationexecutor/operation_generator.go]: Because if the func returns not nil, it will trigger exponentialBackOffError. Thus, when `mount` function will be called is largely decided by how log `attach` function will last. If `attach` function lasts 8s, then, the `mount` function(0.5 +1 +2 +4 < 8, retry for 4 times) is not ready and it would be called 8s later. If `attach` function lasts 16s, attach function would be called 16s later. it is Intolerable for users.

**Comment by cwdsuzhou** [comment]
cc @jsafrane @msau42 @gnufied

**Comment by jsafrane** [comment]
If I remember correctly, we need the volume in node.status.volumesInUse to prevent races in A/D controller. If we allow mounting of volumes that are not in  `volumesInUse`, then A/D controller has no way how to check that volume is not mounted / not being mounted before the controller detaches it.

Correct way would be to fix kubelet to update node.status more often / on VolumeManager request. @gnufied, @jingxu97, was there an issue for this?

**Comment by cwdsuzhou** [comment]
> If I remember correctly, we need the volume in node.status.volumesInUse to prevent races in A/D controller. If we allow mounting of volumes that are not in `volumesInUse`, then A/D controller has no way how to check that volume is not mounted / not being mounted before the controller detaches it.
> 
> Correct way would be to fix kubelet to update node.status more often / on VolumeManager request. @gnufied, @jingxu97, was there an issue for this?

I think it is necessary to check if the volumes are in `volumesInUse`, however if the function returns a not-nil value, it may cause ExponentialBackOffError. Return nil when `ReportedInUse` is false seems would not bring extra costs and would not influence the `detach` as you said.  What's more, though `ReportedInUse` is true, the desired volume may still not be found on the node(heartbeat may has not reported). So, the function would better to return nil if the volume is not found.

**Comment by gnufied** [comment]
The reason we do not proceed with mount if volume is not reported in-use is because attach/detach controller otherwise has no way of knowing if a volume is in use and it may detach a volume that is being mounted (if mount is allowed to proceed without reporting as in-use).

I am working on https://github.com/kubernetes/kubernetes/pull/75213 which should speed up some of these operations because then exponential backoff of one operation will not affect other.  

We could also update node's status more often but it has to be verified by node sig because that has performance implications.

**Comment by jingxu97** [comment]
I think this fix should be ok. It is still checks volumesInUse, but change it to if volumeInUse has not yet updated, do nothing, and reconciler will come back and check it again. 

As @cwdsuzhou mentioned, the current behavior is that if volumeInUse check fails (node status has not yet updated), it will returns error so that when reconciler tries to check it again, it will cause exponential backoff and delay the following checks. With this fix, reconciler can recheck volume is attach faster without delay. Just need to make sure we already have tests to cover this.

**Comment by gnufied** [comment]
Yeah I think while the fix may be okay but I am not sure about returning nil from both the places. This pretty much would disable exponential backoff mechanism for this operation and in a large enough cluster could result in increased API usage from *all* nodes.

**Comment by jingxu97** [comment]
right, expoential backoff is too slow, but reconcile every 100 ms seems too fast. Could we have a normal backoff (e.g, every 2s :)

**Comment by cwdsuzhou** [comment]
Yes, it would result in creased API usage. However,  backoff is currently set for all operations, directly change it may lead to more issues. Adding a normal backoff as @jingxu97 mentioned seems to be a good idea in the feature. I think adding a variable to limit the call frequency my be a simple and effective way Now.

**Comment by cwdsuzhou** [comment]
@jingxu97 @gnufied , to avoid the potential increased api usage, I add a map to record the last verify time and limit the request frequency temporarily. However, I think we can implement a normal backoff to fix it in the feature.

**Comment by gnufied** [comment]
I am not sure about building wheels within wheels to get around exponential backoff. It makes code hard to follow and reason about.

**Comment by cwdsuzhou** [comment]
yes,  I think it is only a temporary solution to fix the issue. Do you have a better way now ?@gnufied

**Comment by jingxu97** [comment]
actually VolumeInUse field is updated in node status by kubelet periodically (every 10 seconds). So no matter how fast we check this field through VerifyControllerAttachedVolume, it will not increase API calls. So I think the change before in this PR can work, it makes sure to pass the check VolumeInUse as long as it is updated in the node status without much delay. It will not help to speed up kubelet to update the node status. But I think 10 seconds seems reasonable considering the normal volume attach time.

**Comment by gnufied** [comment]
Yeah, that is why I think returning nil from *both* places is the part that worries me bit - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L1348.  Returning nil from first place is probably fine (from API perspective) but there is still a cost associated with spinning new goroutines too quickly. 

Returning nil from second place was problematic because even though volume is marked as in-use it may not be attached yet.

File: pkg/volume/util/operationexecutor/operation_generator.go [hunk 2]
```diff
@@ -1344,8 +1344,8 @@ func (og *operationGenerator) GenerateVerifyControllerAttachedVolumeFunc(
 			}
 		}
 
-		// Volume not attached, return error. Caller will log and retry.
-		return volumeToMount.GenerateError("Volume not attached according to node status", nil)
+		// Volume not attached, do not return err. Kubelet will reconcile the actual and desired status.
+		return nil, nil
```
Comment (gnufied) [inline pkg/volume/util/operationexecutor/operation_generator.go]: @jingxu97 this is the return that would cause API usage.

**Comment by msau42** [comment]
> actually VolumeInUse field is updated in node status by kubelet periodically (every 10 seconds).

is this true even with the NodeLease feature?  I had to fix a bug around that in 1.14.

File: pkg/volume/util/operationexecutor/operation_generator.go (hunk 2 above)
Comment (cwdsuzhou) [reply pkg/volume/util/operationexecutor/operation_generator.go]: If return not nil, exponential backoff may cause the volume can not be mounted in time. The performance of attach fucnction is not sure. If the performance is very poor, it may cost very long. However, I think when Kebelet calls the mount function should not be affected by it.

**Comment by cwdsuzhou** [comment]
> > actually VolumeInUse field is updated in node status by kubelet periodically (every 10 seconds).
> 
> is this true even with the NodeLease feature? I had to fix a bug around that in 1.14.

I am not sure if it exists in latest 1.14-beta2, but it exists in  1.14-beta1.

File: pkg/volume/util/operationexecutor/operation_generator.go (hunk 2 above)
Comment (jingxu97) [reply pkg/volume/util/operationexecutor/operation_generator.go]: oh, sorry I missed this change. Can you just make the first change (return nil while checking volumeInUse) and see how much performance is improved? 

If we need further performance enhancement, there are another option we can consider: Decrease the maximum value of exponential backoff. Currently I think we set it to 2 mins, I think reduce to 15-30 seconds seems fine.

One more parameter I have been thinking to adjust is the reconciler looptime. 100ms seems too short, we end up having a lot go routine and just found there are existing operations pending.

**Comment by gnufied** [comment]
I think k8s can't possibly do one-size-fits-all solution here.  If your requirement is attach and mount as soon as possible at the cost of higher API usage, then it may be could be solved by configurable upper bound of exponential backoff. Does making 2min upper bound configurable to 10s or something closer solves your problem?

File: pkg/volume/util/operationexecutor/operation_generator.go (hunk 2 above)
Comment (cwdsuzhou) [reply pkg/volume/util/operationexecutor/operation_generator.go]: @jingxu97 Thank you for your advice. I have removed the 2rd place. I would observe the performance enhancement later.

> If we need further performance enhancement, there are another option we can consider: Decrease the maximum value of exponential backoff. Currently I think we set it to 2 mins, I think reduce to 15-30 seconds seems fine.
> 
> One more parameter I have been thinking to adjust is the reconciler looptime. 100ms seems too short, we end up having a lot go routine and just found there are existing operations pending.

I think it would be wonderful if exponential backoff for `verify` function can be set differently

**Comment by cwdsuzhou** [comment]
> I think k8s can't possibly do one-size-fits-all solution here. If your requirement is attach and mount as soon as possible at the cost of higher API usage, then it may be could be solved by configurable upper bound of exponential backoff. Does making 2min upper bound configurable to 10s or something closer solves your problem?

It may solve the problem.  However, this may increase the api usage from other operations. I think it would be wonderful if exponential backoff for `verify` function can be set differently. In the latest version of this pr, I have removed the 2rd changed place you mentioned.

**Comment by jingxu97** [comment]
@cwdsuzhou did you test the change, could you share some performance change if it is available? Thanks!

**Comment by cwdsuzhou** [comment]
> @cwdsuzhou did you test the change, could you share some performance change if it is available? Thanks!

@jingxu97 
Currently, when I use the original code of the  1.14 beta2, durationBeforeRetry usually reach 8s, seldom reach 4s.
When I return nil in the func if volume has not been reported in use, durationBeforeRetry usually reach 4s or 2s. 
The statistics may exist deviation due to kubelet's report frequency. And it is mainly depended on the performance of `Attach` func in our csi-driver.

**Comment by jingxu97** [comment]
@cwdsuzhou thanks for the information. One more thing, with this change, reconciler will issue mount operation a lot more frequently. Did you notice any overhead of this? whether it is printing out a lot more logs? Thanks!

**Comment by cwdsuzhou** [comment]
Verify  func was called more often before the volume was reported in use, and these logs were prited a lot more.@jingxu97

**Comment by jingxu97** [comment]
@cwdsuzhou can you copy the logs here? we need to consider reduce the logs if necessary

**Comment by cwdsuzhou** [comment]
image
@jingxu97

**Comment by jingxu97** [comment]
This change lgtm, but just want to make sure other reviewers are ok with it.
cc @gnufied

**Comment by msau42** [comment]
/hold
I think we need to do more extensive stress testing and resource usage profiling before considering removing exponential backoff.  This has the potential to signficantly increase cpu usage of kubelet and cause node instability.

**Comment by jingxu97** [comment]
@cwdsuzhou You mentioned there is no obvious cpu overhead change. Please share your observation of cpu/memory usage with deployment a number of volumes if it is available so that we can be confident about this change.

**Comment by jingxu97** [comment]
Comparing two options mentioned here,
1. return nil for volumeInUse check to avoid exponential back off
2. reduce maximum exponential back off time

Both options have procs and cons. Normally it will take about 1-10 seconds to have volumeInUse updated in node status. With option 1, the checking frequency of 100ms (reonciler loop time) is kind of too high. We need to check the overhead caused by it. With option 2, if we could change to maximum of exponential backoff time to about 15s, it will has less overhead compared to 1, but still have some delay for the check. If you could so some experiments on different options, that would be helpful!

**Comment by cwdsuzhou** [comment]
@jinxu97 @msau42 Thank you, I would do more stress testing.

**Comment by gnufied** [comment]
Another option is to expose exponential backoff params to kubelet as a configurable parameter. In some situation where low latency mount is expected, one can configure this to be lower. Obviously there will be tradeoffs but at least it will be user configurable.

**Comment by cwdsuzhou** [comment]
For verifying the option 1 as @jingxu97 mentioned. I set up a single-node 1.14-rc1 cluster and use hostpath csi driver as the test driver. 
50 pv and pvcs were created. Then, I ran `cat pod-temp.yaml |sed -e s/{idx}/${i}/g|kubectl apply -f -` to create pods(10,20,...,50). Results in my experiments are as below.

Avg: avg usage in a minute
Max: max usage in a minute

Changed: return nil for volumeInUse check 
Original: return not nil for volumeInUse check 

Node config: 4 core + 8GB 
Base cpu usage: About 2.3%


CPU Usage Info

  | Avg %(Changed) | Max %(Changed) | Avg %(Original) | Max %(Original)
-- | -- | -- | -- | --
10(pvs) | 7.26 | 10.43 | 6.25 | 10.84
20 | 10.62 | 17.88 | 9.52 | 16.84
30 | 15.40 | 22.22 | 15.24 | 19.41
40 | 18.05 | 23.1 | 17.03 | 20.80
50 | 18.46 | 29.77 | 19.54 | 24.75


Memory increase  can not be obviously observed.  The statistics may exist deviation due to kubelet's report frequency and cpu collecting frequency.

@jingxu97 @msau42 @gnufied

**Comment by jingxu97** [comment]
@cwdsuzhou thanks a lot for sharing the data. Looks like it has a small overhead ranging from 1% - 15% on average from the data. But one strange thing is for for 50 PVCs, the average CPU usage is even smaller than original. I am not sure how many runs you did for collecting the data. So I think there is some concern for overhead. Do you want to try to test the option 2 to reduce the maximum waiting time? I think 2 mins is unnecessary long and 30 seconds or a little bit less are better. It will impact other volume operations e.g. increase the frequency of retrying attach/detach if they fail, but I think setting to retry after 15-30 seconds should not have big impact. You can make a similar test for that too.

**Comment by cwdsuzhou** [comment]
> @cwdsuzhou thanks a lot for sharing the data. Looks like it has a small overhead ranging from 1% - 15% on average from the data. But one strange thing is for for 50 PVCs, the average CPU usage is even smaller than original. I am not sure how many runs you did for collecting the data. So I think there is some concern for overhead. Do you want to try to test the option 2 to reduce the maximum waiting time? I think 2 mins is unnecessary long and 30 seconds or a little bit less are better. It will impact other volume operations e.g. increase the frequency of retrying attach/detach if they fail, but I think setting to retry after 15-30 seconds should not have big impact. You can make a similar test for that too.

Deviation may exist due to our cpu performance collector. And the samples in a minute the tool collector would also influent the results.

Option 2 may work for some scenes. However, it  may not improve the `attach` performance of our own csi driver for that backoff  limit usually reaches 8s , less than 15-30s you mentioned.  Exposing exponential backoff params to kubelet as a configurable parameter, as @gnufied mentioned, seems to be better.

**Comment by jingxu97** [comment]
I mention 30s for a safer setting. Sure, make it configurable is even better. But I think the default value should be set to something less than 2 mins too. To avoid unexpected behavior by setting a different value, some stress timing is needed too.

**Comment by cwdsuzhou** [comment]
> I mention 30s for a safer setting. Sure, make it configurable is even better. But I think the default value should be set to something less than 2 mins too. To avoid unexpected behavior by setting a different value, some stress timing is needed too.

Ok, I will test ty2 later and do some stress testing.  I would also try to expose the exponential backoff params in another PR.

**Comment by cwdsuzhou** [comment]
@jingxu97  I have done some tests. Maximum exponential back off time was set from 5s to 30s(5,10,15,20,25,30),   I have not found an obvious relationship between the backoff value and cpu increase. Detailed data can be found here. https://gist.github.com/cwdsuzhou/4f1e90e98f90cd0a5e129a946c7936d8

**Comment by cwdsuzhou** [comment]
I would like to expose the exponential backoff params in another PR.

**Comment by cwdsuzhou** [comment]
ping @jingxu97 @gnufied

PTAL

**Comment by jingxu97** [comment]
cc @msau42 @saad-ali 

It will increase the frequency of some API calls, but I think API server should handle well with 30s duration. Say if there are 100 volumes attached to the node, then it is about 3 API calls per second?

**Comment by gnufied** [comment]
@jingxu97 The problem is when things go horribly wrong and you are just trying to bring back a badly behaving cluster. In such cases - hitting into API quota issues just exacerbate the problem and since cluster is in bad shapr, everything is trying to reconcile. We have to be super careful lowering this number since I have not seen major complaints about exp. backoff being too high.

**Comment by gnufied** [comment]
Lowering this value afaict  affects all operations, not just operations on a node. It affects attach/detach/resize/provisioning literally everything.

**Comment by cwdsuzhou** [comment]
As @gnufied mentioned, lowering the maxDuration may cause some potential problems. I would close this PR.
I have opened another PR  https://github.com/kubernetes/kubernetes/pull/75963, which exposes a parameter to fix the issue. 

@gnufied @jingxu97 @msau42 PTAL   thanks
